src/datautils/wellboreItemShapeGenerator.ts

```diff
@@ -57,56 +52,93 @@ export const cementDiameterChangeDepths = (
-      outerDiameter = holeAtDepth.diameter - HOLE_OUTLINE;
-    } else {
-      outerDiameter = defaultCementWidth;
-    }
+export const findCementOuterDiameterAtDepth = (innerCasing: Casing[], nonAttachedCasings: Casing[], holes: HoleSize[], depth: number): number => {
+  const defaultCementWidth = 100; // Default to flow cement outside to show error in data
+
+  const innerCasingAtDepth = innerCasing.find((casing) => casing.start <= depth && casing.end >= depth);
+  const innerDiameter = innerCasingAtDepth ? innerCasingAtDepth.diameter : 0;
+
+  const outerCasings = nonAttachedCasings.filter((casing) => casing.innerDiameter > innerDiameter);
+  const holeAtDepth = holes.find((hole) => hole.start <= depth && hole.end >= depth && hole.diameter > innerDiameter);
+  const outerCasingAtDepth = outerCasings
+    .filter((d) => d)
+    .sort((a, b) => a.innerDiameter - b.innerDiameter) // ascending
+    .find((casing) => casing.start <= depth && casing.end >= depth && casing.diameter > innerDiameter);
+
+  let outerDiameter;
+  if (outerCasingAtDepth) {
+    outerDiameter = outerCasingAtDepth.innerDiameter;
+  } else if (holeAtDepth) {
+    outerDiameter = holeAtDepth.diameter - HOLE_OUTLINE;
+  } else {
+    outerDiameter = defaultCementWidth;
+  }
+
+  return outerDiameter;
+};
+
+export const createComplexRopeSegmentsForCement = (
+  cement: Cement,
+  casings: Casing[],
+  holes: HoleSize[],
+  exaggerationFactor: number,
+  getPoints: (start: number, end: number, interestPoints: number[]) => MDPoint[],
+): ComplexRopeSegment[] => {
+  // Merge deprecated casingId and casingIds array
+  // TODO remove casingId now?
+  const casingIds = [cement.casingId, ...(cement.casingIds || [])].filter((id) => id);
+
+  const attachedCasings = casingIds.map((casingId) => casings.find((casing) => casing.casingId === casingId));
+  if (attachedCasings.length === 0 || attachedCasings.includes(undefined)) {
+    throw new Error('Invalid cement data, cement is missing attached casing');
+  }
+
+  const topOfCement = cement.toc;
+  attachedCasings.sort((a: Casing, b: Casing) => a.end - b.end); // ascending
+  const bottomOfCement = attachedCasings[attachedCasings.length - 1].end;
+
+  const { outerCasings, overlappingHoles } = findIntersectingItems(topOfCement, bottomOfCement, attachedCasings, casings, holes);
+
+  const outerDiameterIntervals = [...outerCasings, ...overlappingHoles].map((d) => ({
+    start: d.start,
+    end: d.end,
+  }));
+
+  const changeDepths = cementDiameterChangeDepths(cement, bottomOfCement, outerDiameterIntervals);
+
+  const diameterIntervals = changeDepths
+    .map((depth, index, list) => {
+      if (index === 0) {
+        return;
+      }
+
+      const nextDepth = list[index - 1];
+      const diameter = findCementOuterDiameterAtDepth(attachedCasings, outerCasings, overlappingHoles, depth) * exaggerationFactor;
+
+      return { top: nextDepth, bottom: depth, diameter };
+    })
+    .filter((d) => d);
```

Comment:
Could we utilize `reduce` or `flatMap` here instead?

Reply:
to avoid the filter call?

Reply:
Yes, as it is caused by L123.